### PR TITLE
kernel: Add dynamic interrupt priority API

### PR DESCRIPTION
**Pull request by besmarsh**
This PR introduces a new API `irq_priority_set()`, which enables setting interrupt priority dynamically.

Existing internal architecture-specific IRQ priority setting functionality is now exposed through `arch_irq_priority_set()`, and architectures that do not support this have stubs to prevent build errors.

Testing of this API has been added to the existing nested_irq kernel interrupt test as it naturally can be very easily tested by that existing test.

### CODE REVIEW

**Comment by evgeniy-paltsev** [comment]
Hi, I'm wondering about the rational of such API? Is there any real scenario when we need to set priority not once on the init but in the runtime?

If we do want to make this API something standard we should think more about the corner cases.

For example, on ARC SMP target that operation isn't atomic (after been properly implemented) as we need to set interrupt priorities on each core private interrupt controller. That basically means we need to do inter-core interrupts to make this setup. 

So doing that such interrupt priority setup is not the thing which is safe when the interrupt line is enabled and interrupt may arrive.

IMO, at least we should add retirement that interrupt line should be disabled when we do dynamic priority setup.

File: include/zephyr/sys/arch_interface.h [hunk 1]
```diff
@@ -307,6 +307,16 @@ int arch_irq_connect_dynamic(unsigned int irq, unsigned int priority,
...
 			     const void *parameter, uint32_t flags);
 
+/**
+ * Arch-specific hook to set an interrupt's priority
+ *
+ * @param irq IRQ line number
+ * @param priority Interrupt priority
+ * @param flags Arch-specific IRQ configuration flag
+ */
+void arch_irq_priority_set(unsigned int irq, unsigned int priority,
```
Comment (evgeniy-paltsev) [inline include/zephyr/sys/arch_interface.h:317]: Hi, I'm wondering about the rational of such API? Is there any real scenario when we need to set priority not once on the init but in the runtime?

If we do want to make this API something standard we should think more about the corner cases.

For example, on ARC SMP target that operation isn't atomic (after been properly implemented) as we need to set interrupt priorities on each core private interrupt controller. That basically means we need to do inter-core interrupts to make this setup. 

So doing that such interrupt priority setup is not the thing which is safe when the interrupt line is enabled and interrupt may arrive.

IMO, at least we should add retirement that interrupt line should be disabled when we do dynamic priority setup.

**Review by evgeniy-paltsev — CHANGES_REQUESTED** [review]

File: include/zephyr/sys/arch_interface.h (hunk 1 above)
Comment (besmarsh) [reply include/zephyr/sys/arch_interface.h:317]: Hi, thanks for your comment. I think there are use-cases for priority changing during the runtime of an application. Priority is set at runtime anyway (even if during init) so this change just allows it to be set by the application at any point. Also, `irq_connect_dynamic()` already allows runtime interrupt priority setting, this new API does the same but without configuring the ISR.

You make a good point that this operation could be unsafe and taking care such as disabling the interrupt line while priority is changed would be sensible. I think this should be handled by each architecture's implementation - I had assumed that as `irq_connect_dynamic()` already configures interrupt priority at runtime that architectures should already have safe implementations. Note that this PR doesn't add any architecture-specific implementations, it just uses already existing ones.

Comment (evgeniy-paltsev) [reply include/zephyr/sys/arch_interface.h:317]: > I think there are use-cases for priority changing during the runtime of an application.

So what is the reason for adding the API if we don't even have an idea how we can benefit from it? :)

>> so this change just allows it to be set by the application at any point.
>
>> Note that this PR doesn't add any architecture-specific implementations, it just uses already existing ones.

That is completely not true! Currently 'dynamic interrupt priority setup' has multiple restrictions which we break by this new API.

Currently the dynamic interrupt setup may be done only **once** during the system execution as `irq_connect_dynamic` may be called only once per interrupt line. It's clearly specified in the documentation, see:  

```
Removing or re-configuring a dynamic interrupt is currently unsupported.
```

Additional conclusion from this is that dynamic priority setup can be done only on disabled interrupt line - as we can't enable line if we don't instantiate IRQ handler (with `irq_connect_dynamic`).

And there is deep rational behind all that limitations - we can't just change interrupt priority on the enabled line without IRQ handling code overhaul (or at least verify) on all architectures - as currently we don't expect that we may enter the interrupt with one priority, but exit with it's priority changed. Please don't forget that Zephyr supports multi-core systems and it's perfectly valid corner case! (when we process IRQ on one core and change this IRQ priority from another core).

Moreover such scenario may not be supported by some interrupt controllers HW itself.

---

In conclusion. We try to introduce the API:

- we don't have clear vision about how and why it can / should be used
- which breaks current assumptions (and we don't verify if it is actually valid without such assumptions)
- we don't sure if it can properly implemented on all architectures

I'd propose to avoid that.

Comment (besmarsh) [reply include/zephyr/sys/arch_interface.h:317]: > So what is the reason for adding the API if we don't even have an idea how we can benefit from it? :)

I wouldn't have requested the feature if I had no use for it :) My bad, I should have explained my use-cases. Consider any application that has multiple operating modes/states, and between these different modes/states the relative priorities of interrupt-driven functionality change. Also, this would overcome a limitation of setting interrupt priority statically through devicetree where (please correct me if I'm wrong) to set the interrupt priority of a device you must also set the IRQ line even though this is already set and will not be changed.

> Currently the dynamic interrupt setup may be done only once during the system execution as `irq_connect_dynamic` may be called only once per interrupt line.

> Additional conclusion from this is that dynamic priority setup can be done only on disabled interrupt line - as we can't enable line if we don't instantiate IRQ handler (with `irq_connect_dynamic`).

Okay, yes, I hadn't appreciated that the use of `irq_connect_dynamic` relied upon the interrupt line already being disabled. Would there be a problem with disabling the interrupt line prior to changing the interrupt priority to make this a safe operation? That wouldn't be hard to add, but if that's not sufficient then yes this feature does become more problematic.

> we don't sure if it can properly implemented on all architectures

I don't think that not being supported on all architectures/HW should stop the feature being introduced, interrupt priority is not supported on all archs/HW and of course that is supported by Zephyr.

Comment (tbursztyka) [reply include/zephyr/sys/arch_interface.h:317]: > Would there be a problem with disabling the interrupt line prior to changing the interrupt priority to make this a safe operation?

Depends on the interrupt controller. I have x86 IOAPIC in mind but I guess this might be the case elsewhere too. You would need to have a migration procedure: where old setup would need a proper closure (make sure pending irq are handled etc...), disabling, then de-allocate any related object, reallocate, enabling... 

Imo, if such feature is required, it will be nice to see it properly implemented for all relevant arch instead of having dummy functions doing nothing.

**Review by peter-mitsis — COMMENTED** [review]
There may be some hardware specific subtleties that I am missing, but this generally looks good to me. However, I would recommend one thing--in the updated documentation, add some notes about potential issues using this at run time (such as those previously identified), and best practices on how to best use it safely.

**Review by andyross — CHANGES_REQUESTED** [review]
I'm not opposed in principle to promoting the ARM IRQ set API to a OS API, but can we do this in a less boilerplatey way?  Many (most) hardware architectures can't actually do this at all.  Xtensa can't, IRQ priorities are hardwired at synthesis time.  x86 can't, priority is encoded in the interrupt vector instead.    Posix doesn't even have preemption of any form, so it couldn't use this.  Don't think MIPS has anything standard that does this.  Not sure about ARC or NIOS.  But all those architectures have a bunch of changes anyway?

Can you wrap this behind a CONFIG_ARCH_HAS_WHATEVER kind of thing and provide fallbacks in the core headers instead?

Conversely, up a level, some interrupt controllers do provide something like this.  But... this patch doesn't have a framework for "driver" level integration, only arch.  Seems like that deserves some thought too.

**Comment by andyross** [comment]
(Pfft, I meant RISC-V but typed "MIPS", which tells you something about how that architecture is filed away in my brain, I guess.)

**Comment by besmarsh** [comment]
> I'm not opposed in principle to promoting the ARM IRQ set API to a OS API, but can we do this in a less boilerplatey way? Many (most) hardware architectures can't actually do this at all. Xtensa can't, IRQ priorities are hardwired at synthesis time. x86 can't, priority is encoded in the interrupt vector instead. Posix doesn't even have preemption of any form, so it couldn't use this. Don't think MIPS has anything standard that does this. Not sure about ARC or NIOS. But all those architectures have a bunch of changes anyway?
> 
> Can you wrap this behind a CONFIG_ARCH_HAS_WHATEVER kind of thing and provide fallbacks in the core headers instead?

@andyross Would you prefer to have the feature wrapped in something like `CONFIG_ARCH_HAS_DYNAMIC_IRQ_PRIO_SET`, and when this isn't set the function could return `-ENOTSUP` or similar? That would avoid having to have changes in all the architectures that don't support the feature.

> Conversely, up a level, some interrupt controllers do provide something like this. But... this patch doesn't have a framework for "driver" level integration, only arch. Seems like that deserves some thought too.

I'm not sure how would be best to go about allowing interrupt controllers to provide the functionality. Perhaps another config (e.g. `CONFIG_INTERRUPT_CONTROLLER_HAS_DYNAMIC_IRQ_PRIO_SET`) could be used to facilitate this?

**Comment by andyross** [comment]
> @andyross Would you prefer

I'm not picky about naming or microdesign (though maybe someone else has strong feelings), I just want the overhead to be borne by only the architectures with the (somewhat obscure) feature and not everyone.

And I don't have specific ideas about how to integrate driver-level interrupt controllers either, which is sort of ad-hoc already.  I was just pointing it out as something to be aware of -- how does this interact with the next level up.

**Comment by carlescufi** [comment]
@andyross @evgeniy-paltsev can you please take another look?

**Comment by carlescufi** [comment]
@besmarsh thanks for rebasing. Please join our Discord and ping me there if this PR doesn't see progress.

**Comment by andyross** [comment]
This is still pretty boilerplatey.  We really can't manage this with something like a kconfig CONFIG_HAS_IRQ_PRIORITY_SET, or a weak symbol, or something?  Seems like a lot of impact to systems that won't ever be able to support the feature.

Alternatively, if that's really not possible for some reason: maybe some work done to enable a runtime assert (or better: a build time failure) on systems without support?

Just to repeat from before: this looks like a portable cross-platform API, but it's exposing a feature that is absolutely not portable at all!  That doesn't mean we shouldn't provide it, but we need to do it in a way that doesn't make things a mess for the systems that can't do it.

**Comment by besmarsh** [comment]
Thanks for checking in on this guys. I rebased it just to keep it alive. I know I need to take another look and make some changes, I just haven't had time to do so recently.

Don't worry about any further reviewing until I get some time to look at making the suggested changes.

**Comment by carlocaione** [comment]
> Thanks for checking in on this guys. I rebased it just to keep it alive. I know I need to take another look and make some changes, I just haven't had time to do so recently.

It's probably better if you make this a draft then, otherwise people will get confused.
